source/extensions/filters/http/header_mutation/header_mutation.cc

```diff
@@ -42,26 +41,15 @@ Http::FilterHeadersStatus HeaderMutation::decodeHeaders(Http::RequestHeaderMap&
       Http::Utility::resolveMostSpecificPerFilterConfig<PerRouteHeaderMutation>(decoder_callbacks_);
 
   if (route_config_ != nullptr) {
-    route_config_->mutations().mutateRequestHeaders(headers, decoder_callbacks_->streamInfo());
+    route_config_->mutations().mutateRequestHeaders(headers, ctx, decoder_callbacks_->streamInfo());
   }
 
   return Http::FilterHeadersStatus::Continue;
 }
 
 Http::FilterHeadersStatus HeaderMutation::encodeHeaders(Http::ResponseHeaderMap& headers, bool) {
-  // There is an corner case that the downstream request headers will nullptr when the request is
-  // reset (for example, reset by the stream idle timer) before the request headers are completely
-  // received. The filter chain will be created and the encodeHeaders() will be called but the
-  // downstream request headers will be nullptr.
-  const Http::RequestHeaderMap* downstream_request_headers =
-      encoder_callbacks_->streamInfo().getRequestHeaders();
-  const Http::RequestHeaderMap& request_headers =
-      downstream_request_headers != nullptr
-          ? *downstream_request_headers
-          : *Http::StaticEmptyHeaders::get().request_headers.get();
-
-  config_->mutations().mutateResponseHeaders(request_headers, headers,
-                                             encoder_callbacks_->streamInfo());
+  Formatter::HttpFormatterContext ctx{encoder_callbacks_->requestHeaders().ptr(), &headers};
```

Comment:
Should we worry about described corner case? I believe that if that situation happens, `encoder_callbacks_->requestHeaders().ptr()` will be nullptr and the context will handle it gracefully, correct?

Reply:
Yeah.